distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/PluginsManifest.java

```diff
@@ -0,0 +1,151 @@
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * This class models the contents of the {@code elasticsearch-plugins.yml} file. This file specifies all the plugins
+ * that ought to be installed in an Elasticsearch instance, and where to find them if they are not an official
+ * Elasticsearch plugin.
+ */
+public class PluginsManifest {
+    private final List<PluginDescriptor> plugins;
+    private final String proxy;
+
+    @JsonCreator
+    public PluginsManifest(@JsonProperty("plugins") List<PluginDescriptor> plugins, @JsonProperty("proxy") String proxy) {
+        this.plugins = plugins == null ? List.of() : plugins;
+        this.proxy = proxy;
+    }
+
+    /**
+     * Validate this instance. For example:
+     * <ul>
+     *     <li>All {@link PluginDescriptor}s must have IDs</li>
+     *     <li>Any proxy must be well-formed.</li>
+     *     <li>Unofficial plugins must have URLs</li>
+     * </ul>
+     *
+     * @param manifestPath the path to the file used to create this instance. Used to construct error messages.
+     * @throws UserException if validation problems are found
+     */
+    public void validate(Path manifestPath) throws UserException {
+        if (this.plugins.stream().anyMatch(each -> each == null || each.getId() == null || each.getId().isBlank())) {
+            throw new RuntimeException("Cannot have null or empty plugin IDs in: " + manifestPath);
+        }
+
+        final Set<String> uniquePluginIds = new HashSet<>();
+        for (final PluginDescriptor plugin : plugins) {
+            if (uniquePluginIds.add(plugin.getId()) == false) {
+                throw new UserException(ExitCodes.USAGE, "Duplicate plugin ID [" + plugin.getId() + "] found in: " + manifestPath);
+            }
+        }
+
+        for (PluginDescriptor plugin : this.plugins) {
+            if (InstallPluginAction.OFFICIAL_PLUGINS.contains(plugin.getId()) == false && plugin.getUrl() == null) {
+                throw new UserException(
+                    ExitCodes.CONFIG,
+                    "Must specify URL for non-official plugin [" + plugin.getId() + "] in " + manifestPath
+                );
+            }
+        }
+
+        if (this.proxy != null) {
+            final String[] parts = this.proxy.split(":");
+            if (parts.length != 2) {
+                throw new UserException(ExitCodes.CONFIG, "Malformed [proxy], expected [host:port] in: " + manifestPath);
+            }
+
+            if (ProxyUtils.validateProxy(parts[0], parts[1]) == false) {
+                throw new UserException(ExitCodes.CONFIG, "Malformed [proxy], expected [host:port] in: " + manifestPath);
+            }
+        }
+
+        for (PluginDescriptor p : plugins) {
+            if (p.getUrl() != null) {
+                try {
+                    new URL(p.getUrl());
```

Comment:
We should use `URI` here instead since `new URL()` performs a DNS lookup on hostnames so folks might get errors just due to network funniness.

Reply:
Ooops, I thought I'd already caught all the places I was mis-using `URL` 👍